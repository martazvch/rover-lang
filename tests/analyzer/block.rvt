// Test blocks

// Void assignment

code:
fn main() {
    var noval = {
        var integer = 1
    }

    var void_post_assign
    void_post_assign = {
        _ = 3 + 4
        var empty
    }
}

error:
VoidAssignment
VoidAssignment

================

code:
fn main() {
    {5}
    print "unused"
}

error:
UnusedValue

================

code:
fn main() {
    var a: int = {
        var nb = 7.5
        nb
    }
}

error:
InvalidAssignType, int, float

================

code:
// Declare local 0, declare local 1, use local 0 in binop
// assign to local 1, returns local 0, defines global 0
fn main() {
    var a: int = {
        var nb = 7
        var c = 56
        c = nb - 5
        nb
    }
}

expect:
[Fn declaration, scope Global, index 0, return ImplicitVoid]
[Block pop count 2]
[Variable scope Local, index 1]
[Variable scope Local, index 2]

[Binop cast None, type int]
[Variable scope Local, index 1]

// Assign to "c"
[Variable scope Local, index 2]
// Return "nb"
[Variable scope Local, index 1]

// Assign to "a"
[Variable scope Local, index 1]

================
