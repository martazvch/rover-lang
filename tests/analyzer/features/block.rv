fn main() {
    var a: int = {
        var nb = 7
        var c = 56
        c = nb - 5
        nb
    }
}
