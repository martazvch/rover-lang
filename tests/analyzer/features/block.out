[Fn declaration main, index: 0, scope: global, return kind: implicit_void]
    [Variable declaration index: 1, scope: local]
        [Block pop count: 2, is_expr: true]
            [Variable declaration index: 1, scope: local]
                [Int 7]
            [Variable declaration index: 2, scope: local]
                [Int 56]
            [Assignment index: 2, scope: local]
                [Binop type: SubInt, cast: none]
                    [Variable index: 1, scope: local]
                    [Int 5]
            [Variable index: 1, scope: local]
